### PR TITLE
SoundWire: fix jack detection

### PR DESCRIPTION
**Pull request by plbossart**
Alternate code to PR #4768 to avoid changing the behavior for HDAudio codecs.

### CODE REVIEW

**Review by bardliao — DISMISSED** [review]

**Comment by bardliao** [comment]
@plbossart On second thought, the existing `mask` is used as `val`. Will it be more readable if we set `mask` bit only if a codec is present? And we won't touch the SDW bits by that way.
```
        unsigned int mask = 0;
        unsigned int val = 0;

        if (enable) {
                list_for_each_codec(codec, hbus) {
                        /* only set WAKEEN when needed for HDaudio codecs */
                        mask |= BIT(codec->core.addr);
                        if (codec->jacktbl.used)
                                val |= BIT(codec->core.addr);
                }
        } else {
                list_for_each_codec(codec, hbus) {
                        /* reset WAKEEN only HDaudio codecs */
                        mask |= BIT(codec->core.addr);
                }
        }

        snd_hdac_chip_updatew(bus, WAKEEN, mask & STATESTS_INT_MASK, val);
```

**Comment by plbossart** [comment]
> @plbossart On second thought, the existing `mask` is used as `val`. Will it be more readable if we set `mask` bit only if a codec is present? And we won't touch the SDW bits by that way.
> 
> ```
>         unsigned int mask = 0;
>         unsigned int val = 0;
> 
>         if (enable) {
>                 list_for_each_codec(codec, hbus) {
>                         /* only set WAKEEN when needed for HDaudio codecs */
>                         mask |= BIT(codec->core.addr);
>                         if (codec->jacktbl.used)
>                                 val |= BIT(codec->core.addr);
>                 }
>         } else {
>                 list_for_each_codec(codec, hbus) {
>                         /* reset WAKEEN only HDaudio codecs */
>                         mask |= BIT(codec->core.addr);
>                 }
>         }
> 
>         snd_hdac_chip_updatew(bus, WAKEEN, mask & STATESTS_INT_MASK, val);
> ```

much better indeed, thanks for the suggestion @bardliao

**Comment by plbossart** [comment]
back to draft since the WAKESTS handling needs to be verified.

File: sound/soc/sof/intel/hda-ctrl.c [hunk 1]
```diff
@@ -225,7 +225,7 @@ int hda_dsp_ctrl_init_chip(struct snd_sof_dev *sdev)
 
 	/* clear WAKESTS */
 	snd_sof_dsp_write(sdev, HDA_DSP_HDA_BAR, SOF_HDA_WAKESTS,
-			  SOF_HDA_WAKESTS_INT_MASK);
+			  bus->codec_mask);
```
Comment (bardliao) [inline sound/soc/sof/intel/hda-ctrl.c:228]: Do we want to take the opportunity to clear only SDW codecs WAKESTS in intel_shim_wake()?
I.e.
Change
```
		/* Clear wake status (W1C) */
		wake_sts = snd_hdac_chip_readw(sdw->link_res->hbus, STATESTS);
		wake_sts |= lsdiid;
		snd_hdac_chip_writew(sdw->link_res->hbus, STATESTS, wake_sts);
```
to
```
		snd_hdac_chip_writew(sdw->link_res->hbus, STATESTS, lsdiid);
```

Comment (plbossart) [reply sound/soc/sof/intel/hda-ctrl.c:228]: yep, that sounds good.

Those W1C bits are nasty, when trying to do the right thing with a read-modify-write you end-up adding side effects without realizing it...

Comment (plbossart) [reply sound/soc/sof/intel/hda-ctrl.c:228]: done

**Review by bardliao — DISMISSED** [review]

**Comment by plbossart** [comment]
we'd need to see if this actually works reliably, @aiChaoSONG @keqiaozhang can you help?

**Comment by bardliao** [comment]
> we'd need to see if this actually works reliably, @aiChaoSONG @keqiaozhang can you help?

@plbossart, @keqiaozhang and I did some test, unfortunately, it doesn't work. STATESTS will still read as 0 even we don't clear it.

**Comment by plbossart** [comment]
> > we'd need to see if this actually works reliably, @aiChaoSONG @keqiaozhang can you help?
> 
> @plbossart, @keqiaozhang and I did some test, unfortunately, it doesn't work. STATESTS will still read as 0 even we don't clear it.

Thanks for testing @bardliao

what part of the code do we still see STATESTS as zero? intel_shim_check_wake() ?

**Comment by bardliao** [comment]
> > > we'd need to see if this actually works reliably, @aiChaoSONG @keqiaozhang can you help?
> > 
> > 
> > @plbossart, @keqiaozhang and I did some test, unfortunately, it doesn't work. STATESTS will still read as 0 even we don't clear it.
> 
> Thanks for testing @bardliao
> 
> what part of the code do we still see STATESTS as zero? intel_shim_check_wake() ?

Yes, intel_shim_check_wake().

**Comment by mengdonglin** [comment]
This is latest fix for SoundWire jack detection issue on LNL #4687

**Comment by bardliao** [comment]
> > > > we'd need to see if this actually works reliably, @aiChaoSONG @keqiaozhang can you help?
> > > 
> > > 
> > > @plbossart, @keqiaozhang and I did some test, unfortunately, it doesn't work. STATESTS will still read as 0 even we don't clear it.
> > 
> > 
> > Thanks for testing @bardliao
> > what part of the code do we still see STATESTS as zero? intel_shim_check_wake() ?
> 
> Yes, intel_shim_check_wake().

Looks like we still need [soundwire: intel_ace2x: simplify check_wake() · thesofproject/linux@34d72e3 (github.com)](https://github.com/thesofproject/linux/commit/34d72e3daca37ddd17cd99103ce4e132aaa7f3d7) @plbossart

**Comment by plbossart** [comment]
Re-added the simplification to wake-up all the links in the HDaudio case, please recheck @bardliao @keqiaozhang

**Review by keqiaozhang — APPROVED** [review]
Confirmed that this PR can fix the JD issue https://github.com/thesofproject/linux/issues/4687 on LNL-SDW platform.

**Review by bardliao — APPROVED** [review]

**Comment by plbossart** [comment]
Thanks for the tests, let's merge!

**Pull request merged**